kernel/src/syscall.rs

```diff
@@ -394,12 +398,40 @@ pub enum SyscallReturnVariant {
     FailureU32 = 1,
     FailureU32U32 = 2,
     FailureU64 = 3,
+    FailurePtrUsize = 4,
+    FailurePtrPtr = 5,
     Success = 128,
     SuccessU32 = 129,
     SuccessU32U32 = 130,
     SuccessU64 = 131,
     SuccessU32U32U32 = 132,
     SuccessU32U64 = 133,
+    SuccessUsize = 134,
+    SuccessPtr = 135,
+    SuccessPtrUsize = 136,
+    SuccessPtrPtr = 137,
+}
+
+impl SyscallReturnVariant {
+    /// Maps newly introduced return variants (usize and ptr)
+    /// to old ones (u32) for backwards compatibility.
+    /// This should not be used for any newly designed interfaces,
+    /// and will eventually be deprecated once all interfaces are updated.
+    pub const fn into_compat(self) -> Self {
+        // We only need to be backwards compatible on 32-bit systems
+        let compat = core::mem::size_of::<usize>() == core::mem::size_of::<u32>();
```

Comment:
Why only on 32-bit platforms? This would require having needing different userspace libraries for 32 and 64-bit since libraries generally expect specific return variants.

Reply:
The intent here is not to use any new codes for existing systems (32-bit being a proxy for "legacy"). We can come up with a better compat as that evolves. Possibly this should be a feature.

This would certainly require changes to libtock-c/libtock-rs for 64-bit system. But that was going to happen anyway, because they don't support 64-bit systems. 

However, it does mean they don't need any change on existing 32-bit systems. The intent is allow designing what we would like the interface to look like, without making invasive change. Possibly, on a major release, compat would always be false.

Reply:
I think what @alevy might be saying here is that this is a fine method to use for a 32-bit ABI on a 64-bit system (which certain ARM architectures embrace, for instance).

Reply:
@lschuermann that's much more clever than what I was saying. I think my comment is stale now.